# Super Dense Coding Assignment

##### In this assignment, you will explore the concept of Super Dense Coding using Qiskit. Super Dense Coding is a quantum communication protocol that allows the transmission of classical information using quantum entanglement.

# Part 1: Transmitting 2 Bits Using Super Dense Coding

## Task 1.1: Understanding the Protocol

##### In the Super Dense Coding protocol, Alice and Bob share a maximally entangled two-qubit state (Bell state). Alice can send 2 bits of classical information to Bob by applying a specific quantum gate based on her message and sending her qubit to Bob. Bob can then decode the message by applying a set of operations on the received qubit.

##### 1.Describe the Super Dense Coding protocol in your own words.
##### 2.Write the steps involved in the protocol.
# Task 1.2: Implementing the Protocol in Qiskit
##### Implement the Super Dense Coding protocol in Qiskit. Specifically, code a function that takes a 2-bit message as input and returns the corresponding quantum circuit.

In [2]:
!pip install qiskit-aer
from qiskit_aer import AerSimulator

Defaulting to user installation because normal site-packages is not writeable
  Obtaining dependency information for qiskit-aer from https://files.pythonhosted.org/packages/f8/bc/4caef52bf661984f25fac76ea7fcc84ecff3667931aca8c0b228fc0411f9/qiskit_aer-0.14.2-cp311-cp311-win_amd64.whl.metadata
   ---------------------------------------- 0.0/9.5 MB ? eta -:--:--
   -- ------------------------------------- 0.5/9.5 MB 15.4 MB/s eta 0:00:01
   --- ------------------------------------ 0.8/9.5 MB 10.6 MB/s eta 0:00:01
   ---- ----------------------------------- 1.2/9.5 MB 9.4 MB/s eta 0:00:01
   ------- -------------------------------- 1.7/9.5 MB 10.0 MB/s eta 0:00:01
   -------- ------------------------------- 2.1/9.5 MB 9.7 MB/s eta 0:00:01
   ---------- ----------------------------- 2.6/9.5 MB 9.7 MB/s eta 0:00:01
   ------------ --------------------------- 3.1/9.5 MB 9.7 MB/s eta 0:00:01
   -------------- ------------------------- 3.5/9.5 MB 9.7 MB/s eta 0:00:01
   ---------------- -------

In [5]:
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister, transpile
from qiskit_aer import AerSimulator
from qiskit.visualization import plot_histogram

def prepare_bell_state():
    q = QuantumRegister(2, 'q')
    c = ClassicalRegister(2, 'c')
    qc = QuantumCircuit(q, c)
    qc.h(q[0])  # Apply Hadamard gate to the first qubit
    qc.cx(q[0], q[1])  # Apply CNOT gate
    return qc

def encode_message(qc, message):
    if message == "00":
        pass  # Do nothing, the message is already |00⟩
    elif message == "01":
        qc.x(0)  # Apply X gate to flip the first qubit
    elif message == "10":
        qc.z(0)  # Apply Z gate to change the phase of the first qubit
    elif message == "11":
        qc.x(0)  # Apply X gate to flip the first qubit
        qc.z(0)  # Apply Z gate to change the phase of the first qubit
    
    return qc

def decode_message(qc):
    qc.cx(0, 1)  # Apply CNOT gate
    qc.h(0)  # Apply Hadamard gate to the first qubit
    
    qc.measure([0, 1], [0, 1])  # Measure both qubits
    
    return qc

# Example usage
message = "10"  # 2-bit message to send
qc = prepare_bell_state()
encoded_qc = encode_message(qc, message)
decoded_qc = decode_message(encoded_qc)
print(decoded_qc)

# Simulate the circuit
simulator = AerSimulator()
compiled_circuit = transpile(decoded_qc, simulator)
sim_result = simulator.run(compiled_circuit).result()
counts = sim_result.get_counts()

     ┌───┐     ┌───┐     ┌───┐┌─┐
q_0: ┤ H ├──■──┤ Z ├──■──┤ H ├┤M├
     └───┘┌─┴─┐└───┘┌─┴─┐└┬─┬┘└╥┘
q_1: ─────┤ X ├─────┤ X ├─┤M├──╫─
          └───┘     └───┘ └╥┘  ║ 
c: 2/══════════════════════╩═══╩═
                           1   0 


## Part 2: Transmitting 3 Bits Using a 3-Qubit Entangled State

### Task 2.1: Proposing a 3-Qubit State

##### Now that you have successfully transmitted 2 bits of information, suppose that Alice and Bob share a 3-qubit entangled state with 2 qubits with Alice and 1 qubit with Bob. Their goal is to transmit 3 bits, using those 3 qubits.

###### Describe the 3-qubit entangled state that is shared between Alice and Bob.

### Task 2.2: Designing the Protocol
##### Design a protocol that allows Alice to transmit a 3-bit message by sending her 2 qubits to Bob. Implement this protocol in Qiskit.

In [6]:
from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister, transpile
from qiskit_aer import AerSimulator
from qiskit.visualization import plot_histogram

def create_ghz_state():
    # Create quantum and classical registers with 3 qubits each
    q = QuantumRegister(3, 'q')
    c = ClassicalRegister(3, 'c')
    qc = QuantumCircuit(q, c)
    
    # Prepare a 3-qubit GHZ state
    qc.h(q[0])
    qc.cx(q[0], q[1])
    qc.cx(q[0], q[2])
    
    return qc

def encode_classical_message(qc, message):
    # Apply encoding based on the classical message
    if message == "000":
        pass  # No operation for "000"
    elif message == "001":
        qc.x(0)
    elif message == "010":
        qc.z(0)
    elif message == "011":
        qc.x(0)
        qc.z(0)
    elif message == "100":
        qc.x(1)
    elif message == "101":
        qc.x(1)
        qc.x(0)
    elif message == "110":
        qc.x(1)
        qc.z(0)
    elif message == "111":
        qc.x(1)
        qc.x(0)
        qc.z(0)
    
    return qc

def decode_quantum_state(qc):
    # Apply the decoding operations
    qc.cx(0, 1)
    qc.cx(0, 2)
    qc.h(0)
    
    # Measure all qubits
    qc.measure([0, 1, 2], [0, 1, 2])
    
    return qc

# Example usage
message = "110"  # 3-bit message to encode and send
qc = create_ghz_state()
encoded_qc = encode_classical_message(qc, message)
decoded_qc = decode_quantum_state(encoded_qc)
print(decoded_qc)

# Simulate the circuit
simulator = AerSimulator()
compiled_circuit = transpile(decoded_qc, simulator)
sim_result = simulator.run(compiled_circuit).result()
counts = sim_result.get_counts()

     ┌───┐          ┌───┐          ┌───┐   ┌─┐
q_0: ┤ H ├──■────■──┤ Z ├──■────■──┤ H ├───┤M├
     └───┘┌─┴─┐  │  ├───┤┌─┴─┐  │  └┬─┬┘   └╥┘
q_1: ─────┤ X ├──┼──┤ X ├┤ X ├──┼───┤M├─────╫─
          └───┘┌─┴─┐└───┘└───┘┌─┴─┐ └╥┘ ┌─┐ ║ 
q_2: ──────────┤ X ├──────────┤ X ├──╫──┤M├─╫─
               └───┘          └───┘  ║  └╥┘ ║ 
c: 3/════════════════════════════════╩═══╩══╩═
                                     1   2  0 


## Part 3: Analyzing Limitations
### Task 3.1: Limitations of Transmitting Information
##### Analyze the limitations of the protocol where Bob tries to send more than 2 bits of information using only 1 qubit. Explain why it is not possible.

## Task 3.2: Improving the Protocol
##### Notice that we are sending 2 qubits to transmit 3 bits of information in the above protocol. We can do better by using 2 qubits to send 4 bits of information by repeating the first protocol twice.

##### Can you think of a way to achieve this, in a similar line to the previous methods, but with one common shared state (and not two separate Bell states)?